persist: implementation of state pubsub over grpc

pH14 opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

Implementation of the Persist PubSub design covered [in this design doc](https://github.com/MaterializeInc/materialize/pull/18894).

With this change, we will:

* Always spin up a new grpc service in `environmentd`
* Conditionally connect to this service from `clusterd` (defaults off, dynamically toggleable via feature flag)
* Conditionally push diffs over established connections (defaults on, but requires the above flag to take effect)

This change will have a corresponding Cloud PR to create the `persist-pubsub` service and pass along the hostname to `environmentd`, though the PR is not blocking.

### Motivation

  * This PR adds a known-desirable feature.

https://github.com/MaterializeInc/materialize/issues/18661 / part of https://github.com/MaterializeInc/materialize/issues/18665 

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -312,21 +382,117 @@ enum StateCacheInit {
+                                .duration_since(SystemTime::UNIX_EPOCH)
+                                .expect("failed to get millis since epoch")
+                                .as_micros();
+                            let latency = now.saturating_sub(timestamp) as f64;
```
i think instead we should convert the message timestamp into a Duration and do the subtraction in "Duration-space"

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -312,21 +382,117 @@ enum StateCacheInit {
+        let cache = Arc::clone(&state_cache);
+        if let Some(mut pubsub_receiver) = pubsub_receiver {
+            let _ = mz_ore::task::spawn(|| "persist::rpc::push_responses", async move {
+                cache.metrics.pubsub_client.receiver.connected.set(1);
```
feels like the whole impl here wants to be pushed into `mod rpc`

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -312,21 +382,117 @@ enum StateCacheInit {
+                                .approx_diff_latency
+                                .observe(latency);
+                        }
+                        ref msg @ None | ref msg @ Some(_) => {
```
nit: what's the story with these `ref`s? my mental model is that's an old pattern matching thing in rust that isn't needed anymore

pH14 replied:
oh this is just so we can pattern match the `None` / `Some(_)`, but take ref to the value rather than moving it so we can `warn!` on it

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -312,21 +382,117 @@ enum StateCacheInit {
+        if let Some(state) = self.get_cached(shard_id) {
+            // WIP: could register callbacks here. if not applied, fallback
+            let applied = state.push_diff(diff);
+            let shard_metrics = self.metrics.shards.shard(shard_id);
```
feels somewhat unfortunate to have to look up the metrics by shard_id, too, because my mind immediately jumps to thinking about race conditions. seems not-crazy to put the ShardMetrics on LockingTypedState and make it available via DynState?

pH14 replied:
done!

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -368,10 +534,18 @@ impl StateCache {
+                                metrics: Arc::clone(&self.metrics),
+                                pubsub_sender: self.pubsub_sender.clone(),
+                            };
                             let state = Arc::new(LockingTypedState {
```
nit: feels like this has gotten complicated enough to pull out into a `LockingTypedState::new` method

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -382,6 +556,9 @@ impl StateCache {
                         // We actually did the init work, don't bother casting back
                         // the type erased and weak version.
+                        if let Some(pubsub_sender) = self.pubsub_sender.as_ref() {
+                            pubsub_sender.subscribe(&shard_id);
```
ah hmm seems ideally that the `PubSubSubscriptionToken` would be something that comes back from a `subscribe` call no?

danhhz commented on src/persist-client/src/cache.rs:
```diff
@@ -344,7 +510,7 @@ impl StateCache {
         loop {
             let init = {
-                let mut states = self.states.lock().await;
+                let mut states = self.states.lock().expect("lock poisoned");
```
oh this can be a sync mutex? nice!

danhhz commented on src/persist-client/src/cfg.rs:
```diff
@@ -176,7 +177,6 @@ impl PersistConfig {
     }
 
     /// Returns a new instance of [PersistConfig] for tests.
-    #[cfg(test)]
```
let's put this back and fix whatever WIP thing caused me to do it in the first place :)

danhhz commented on src/persist-client/src/internal/apply.rs:
```diff
@@ -87,24 +92,28 @@ where
     ) -> Result<Self, Box<CodecMismatch>> {
         let shard_metrics = metrics.shards.shard(&shard_id);
-        let state = shared_states
+        let state = Arc::clone(&shared_states)
```
I think you don't need this Arc::clone?

danhhz commented on src/persist-client/src/internal/apply.rs:
```diff
@@ -541,7 +557,15 @@ where
+    Committed(
+        (
+            SeqNo,
+            VersionedData,
```
the SeqNo and the VersionedData are redundant here

danhhz commented on src/persist-client/src/internal/machine.rs:
```diff
@@ -79,9 +80,18 @@ where
         metrics: Arc<Metrics>,
         state_versions: Arc<StateVersions>,
-        shared_states: &StateCache,
+        shared_states: &Arc<StateCache>,
```
since this unconditionally clones the statecache, let's make it take an owned version and push the clone up a level of the call stack

danhhz commented on src/persist-client/src/internal/service.proto:
```diff
@@ -0,0 +1,40 @@
+        ProtoSubscribe subscribe = 3;
+        ProtoUnsubscribe unsubscribe = 4;
+    }
+
```
nit: extra newline

pH14 replied:
the silliest part is that I did check for this, twice actually, and both times was like, "nope must have been changed somewhere along the way" lol

danhhz commented on src/persist-client/src/internal/service.proto:
```diff
@@ -0,0 +1,40 @@
+}
+
+message ProtoPubSubMessage {
+    bytes timestamp = 1;
```
I think the timestamp is probably better serialized as a u64 seconds and a u64 nanos

danhhz commented on src/persist-client/src/internal/service.rs:
```diff
@@ -0,0 +1,226 @@
+));
+
+#[derive(Debug)]
+pub struct PersistService {
```
I decided partway through prototyping that I think it's cleaner to make this module _only_ generated code and just push all our impls into mod rpc

pH14 replied:
only generated code, as in only

```rust
include!(concat!(
    env!("OUT_DIR"),
    "/mz_persist_client.internal.service.rs"
));
```

and whatever traits we need to implement for those generated types?

danhhz replied:
> whatever traits we need to implement for those generated types?

which traits?

pH14 replied:
was meant more hypothetically, there aren't any traits we need on them right now

danhhz replied:
gotcha. yeah, I was literally thinking the `include!` and nothing else

danhhz replied:
oh and fun fact: rust trait coherence just means you have to impl it in the same _crate_ not necessarily the same _module_

pH14 replied:
sure, that works for me

danhhz commented on src/persist-client/src/internal/service.rs:
```diff
@@ -0,0 +1,226 @@
+    ) -> Result<Response<Self::PubSubStream>, Status> {
+        let root_span = info_span!("persist::push::server");
+        let _guard = root_span.enter();
+        let caller_id = req
```
I lol'd at caller_id 📞

danhhz commented on src/persist-client/src/internal/service.rs:
```diff
@@ -0,0 +1,226 @@
+        // connection and its subscriptions from our shared state when it drops.
+        let metrics = Arc::clone(&self.metrics);
+        tokio::spawn(async move {
+            let root_span = info_span!("persist::push::server_conn");
```
feels like this whole task impl should be moved into its own method. probably ditto for the impls of each message type?

danhhz commented on src/persist-client/src/internal/apply.rs:
```diff
@@ -264,10 +273,15 @@ where
                     self.shard_metrics.cmd_succeeded.inc();
                     self.update_state(new_state);
+                    if let Some(pubsub_sender) = self.pubsub_sender.as_ref() {
+                        if self.cfg.dynamic.pubsub_push_enabled() {
```
I think we probably also want this cfg to short circuit the broadcast in the service running in envd

danhhz reviewed (commented):
first round of comments! I definitely have some questions about errors and retries (which are still WIP here), so there will probably be another round of substantive comments before we get to the pure nitpick passes

pH14 commented on src/persist-client/src/cache.rs:
```diff
@@ -67,23 +77,56 @@ impl Drop for RttLatencyTask {
+    pub async fn new(
+        cfg: PersistConfig,
+        registry: &MetricsRegistry,
+        pubsub: Option<PersistPubSubClientConfig>,
```
there was a really silly reason for moving the client connection into `PersistClientCache::new`, which is that we otherwise don't have access to `Metrics`. Couldn't come up with a satisfying way to make `Metrics` (and potentially `PersistConfig` for feature flag access) accessible at a higher level. We also don't _have_ to use `Metrics` and could instead use a different, more narrowly-scoped struct, but it seems much simpler to me to have everything available on one struct

pH14 commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -711,12 +732,47 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
         PersistConfig::new(&mz_environmentd::BUILD_INFO, now.clone()),
         &metrics_registry,
-    );
+        Some(PersistPubSubClientConfig {
```
the counterpoint is that an `envd` managing 180 shards, 100 of which were broadcast 15x over, used an additional 0.06 CPU

mjibson commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -734,13 +754,45 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
+
+    let persist_pubsub_addr = format!(
+        "http://{}:{}",
+        args.persist_pubsub_hostname,
```
The hostname field is needed to advertise to other clients the routable host for this listener?

pH14 replied:
yep! it'll be provided by the `environment-controller` and likely backed by a Kube Service

pH14 commented:
> Possible to rebase this so it builds on current main? I'd like to try out PR code coverage on this change.

@def- rebased! You're good to go here

pH14 commented on src/persist-client/src/read.rs:
```diff
@@ -901,20 +901,6 @@ where
-
         // The latest state still doesn't have a new frontier for us:
         // watch+sleep in a loop until it does.
         let sleeps = self.metrics.retries.next_listen_batch.stream(
```
open question: do we want to have a different retry policy depending on whether we have pubsub toggled on vs off? or do we just adjust our existing `next_listen_batch_retry_*` knobs + the pubsub knobs in tandem in LD? I was leaning towards the latter because this PR is already massive, but curious to hear thoughts

danhhz replied:
the latter!

pH14 replied:
works for me!

danhhz reviewed (commented):
overall, feels great! the complexity is nicely encapsulated. tests are 💯, bravo!

still have to wrap my head around all the details of mod rpc, but everything I saw there made sense at a scan-level of understanding. checkpointing where I got before the end of my workday in case you wanted to see them, i'll pick the rest of this pass up tomorrow. exciting stuff!

def- commented on src/persist-client/src/cache.rs:
```diff
@@ -288,6 +316,29 @@ where
         self
     }
+
+    fn push_diff(&self, diff: VersionedData) {
```
This function looks interesting to test, seems uncovered

pH14 replied:
ooh I like this code coverage report!

one tricky thing in this case is that we're wrapping this feature behind a flag that defaults off so it's safer to ship / test on individual staging environments. if we ran the tests with `persist_pubsub_client_enabled: true` I'd expect this section to be covered

def- commented on src/persist-client/src/rpc.rs:
```diff
@@ -0,0 +1,1646 @@
+impl proto_persist_pub_sub_server::ProtoPersistPubSub for PersistGrpcPubSubServer {
+    type PubSubStream = Pin<Box<dyn Stream<Item = Result<ProtoPubSubMessage, Status>> + Send>>;
+
+    async fn pub_sub(
```
Can be tested outside of unit tests?

def- reviewed (commented):
I marked some interesting missing coverage based on https://buildkite.com/materialize/coverage/builds/93

teskje commented on src/clusterd/src/bin/clusterd.rs:
```diff
@@ -81,6 +81,7 @@ use anyhow::Context;
 use axum::routing;
 use fail::FailScenario;
 use futures::future;
+use mz_persist_client::rpc::{GrpcPubSubClient, PersistPubSubClient, PersistPubSubClientConfig};
```
Sorry for the drive-by, but this should go into the `mz_` import block below.

pH14 commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -734,13 +754,45 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
-        PersistConfig::new(&mz_environmentd::BUILD_INFO, now.clone()),
-        &metrics_registry,
+
+    let persist_pubsub_addr = format!(
```
`internal_persist_pubsub_listen_addr` will be 127.0.0.1 with the process orchestrator (everything on the same machine) and 0.0.0.0 in Kube so it's available to other pods.

The hostname is separate actually... and maybe I should rename it to `persist-pubsub-url` or `persist-pubsub-service-name` for clarity. In Kube, we'll create a Service `persist-pubsub` that internally routes to `envd`, but gives `clusterd` a stable name to talk to that's decoupled from the lifetime of an `envd` pod, or even having to know that the server is hosted by `envd`

pH14 replied:
so in Kube, we'd expect a Service `persist-pubsub` that routes to `envd` and then to set:

`internal_persist_pubsub_listen_addr: 0.0.0.0:6879` - so that `envd` listens to connections on 6879
`persist-pubsub-hostname: persist-pubsub` - so that `clusterd` knows to connect to `persist-pubsub:6879` (which happens to be served by `envd`) as the server

pH14 commented on src/persist-client/src/cache.rs:
```diff
@@ -314,19 +367,35 @@ enum StateCacheInit {
+    }
+
+    pub(crate) fn apply_diff(&self, shard_id: &ShardId, diff: VersionedData) {
+        if let Some(state) = self.get_cached(shard_id) {
```
in the scramble to get this up I totally forgot this is a mutex call under the hood. yeah lemme poke at this some more

pH14 replied:
okay I came up with an alternative here that's semi-viable, but I think I might leave it for a follow-up PR because it's added complication and I don't think it's necessary while the feature lives behind a flag / is intended for staging. the gist is that we can create a channel/stream of `(ShardId, Weak<dyn DynState>)` from the `StateCache`, which the `subscribe_state_cache_to_pubsub` already has access to, and use it to build a local `HashMap<ShardId, Weak<dyn DynState>>` to avoid any mutex contention.

short-term, as in by tomorrow, I'm leaning towards switching the Mutex to a RwLock and then instrumenting it with some metrics

danhhz replied:
sgtm on all counts

pH14 replied:
okay nvm nerdsnipe very successful, introduced a simpler version than what I mentioned that just fetches a weak ref from the StateCache and holds on to it, so we should be doing only 1 mutex lookup per shard

danhhz commented on src/persist-client/src/rpc.rs:
```diff
@@ -0,0 +1,1646 @@
+        }
+    }
+
+    fn reconnect(&self) {
```
I think a good chunk of these error paths can be exercised by crashing processes (docker containers). @pH14 curious if you can think of any interesting error paths that wouldn't be covered by that?

pH14 replied:
agree, assuming `persist_pubsub_client_enabled: true` is set (which is default off)

we'll also need to update cloudtest to point `clusterd` to the right place

danhhz commented on src/persist-client/src/rpc.rs:
```diff
@@ -0,0 +1,1646 @@
+pub struct ShardSubscriptionToken {
+    pub(crate) shard_id: ShardId,
+    sender: Arc<dyn PubSubSenderInternal>,
+    on_drop_fn: Arc<OnceCell<Box<dyn Fn() + Send + Sync>>>,
```
oh that all does sound unsatisfying, but I think I was asking an easier question, which is "why the double Arc"? I tried poking at this a bit and I didn't immediately see why `on_drop_fn` couldn't omit the Arc wrapper around the OnceCell

pH14 commented on src/controller/src/lib.rs:
```diff
@@ -236,6 +238,9 @@ pub struct Controller<T = mz_repr::Timestamp> {
 
     /// Additional context to pass through to cluster instances.
     pub instance_context: StorageInstanceContext,
+    /// WIP
```
I was very proud of removing every WIP and then merge conflicts added a few back in 😭

pH14 commented on src/sqllogictest/src/runner.rs:
```diff
@@ -902,6 +904,7 @@ impl RunnerInner {
                 postgres_factory: postgres_factory.clone(),
                 metrics_registry: metrics_registry.clone(),
                 scratch_directory: None,
+                persist_pubsub_addr: "http://localhost:6879".into(),
```
no! looks like we bind to a random available port so I'll have to figure out what the right play here is for the process orchestrator

pH14 replied:
nvm this is easy, just ignore it because it's not needed for sqllogictests

pH14 commented on src/persist-client/src/rpc.rs:
```diff
@@ -0,0 +1,1646 @@
+                'reconnect_forever: loop {
+                    metrics.pubsub_client.grpc_connection.connected.set(0);
+
+                    if !dynamic_cfg.pubsub_client_enabled() {
```
that'd be nice, was thinking the same when I wrote this

pH14 commented:
Think I've addressed all of Dan's comments from yesterday. @def- and I will chat tomorrow about ways to test this beyond the existing unit tests

pH14 commented:
@def- the latest commit sets up a Kube Service for `cloudtest` so that the clients can connect to the server hosted on `envd` when prompted to. 

I'm not sure the best way to test it... we need to set the system var `persist_pubsub_client_enabled: true`. Maybe on your fork we pass in `--bootstrap-system-parameter=persist_pubsub_client_enabled=true` to get coverage there?

danhhz reviewed (approved):
i put on my nitpick hat, but i really didn't find much. in particular, the abstractions introduced (both the crate-public mod rpc API as well as the abstractions within mod rpc itself) are fantastic!

assuming @def- is on board, I propose that we target getting this merged in time for the release candidate cut tomorrow. that sets us up to start to turn the feature flag on in staging and get some validation there (i think that would also depend on some cloud work for the new flags and port/service?). in parallel, QA can start work on whatever new test coverage we'd like to get for this, and we wouldn't turn it on in prod until that is finished and they are happy. sound good?

pH14 commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -403,6 +420,17 @@ pub struct Args {
+    /// This URL is passed along with the port specified in `internal_persist_pubsub_listen_addr`
+    /// to `clusterd` for external discovery of the Persist PubSub service hosted within
+    /// `environmentd`.
+    #[clap(
```
I'd strongly prefer if `ExternalHost:Port` were passed in to `environmentd` and I made a pass at it yesterday, but I ran into something tricky with the process orchestrator that I've already paged out somehow. I also think there's a chance I was misunderstanding some code, so lemme take a quick peek at this one again

pH14 replied:
yeah I misunderstood things yesterday, replacing with `persist-pubsub-url` which is effectively `ExternalHost:Port`, and keeping `internal_persist_pubsub_listen_addr` which is going to be `0.0.0.0:Port` or `127.0.0.1:Port`

The ports don't actually need to match, I think it might be nice if our Kube Service maps port 80 to 6879 so all `clusterd` needs to know is `http://persist-pubsub` and they're off and running

danhhz replied:
\o/ yeah that sounds much better

pH14 commented on src/environmentd/tests/util.rs:
```diff
@@ -346,6 +352,7 @@ pub fn start_server(config: Config) -> Result<Server, anyhow::Error> {
             postgres_factory,
             metrics_registry: metrics_registry.clone(),
             scratch_directory: None,
+            persist_pubsub_url: "http://localhost:6879".to_owned(),
```
good spot. I chose to skip the pubsub server for sqllogictests since I don't think it'd get any useful coverage, but spun one up and passed in a real address for testdrive.

cc @def-

pH14 commented:
> assuming @def- is on board, I propose that we target getting this merged in time for the release candidate cut tomorrow. that sets us up to start to turn the feature flag on in staging and get some validation there (i think that would also depend on some cloud work for the new flags and port/service?). in parallel, QA can start work on whatever new test coverage we'd like to get for this, and we wouldn't turn it on in prod until that is finished and they are happy. sound good?

This is also my preference! I think I've addressed all comments now, and setup testdrive and cloudtest to run with a pubsub server / proper discovery (again, assuming `persist_pubsub_client_enabled: true` is set). If this is satisfactory for QA et al, I'm ready to merge this tomorrow so we can start opting in environments on staging next week

def- reviewed (approved):
Fine by me. I triggered another round of test runs: https://buildkite.com/materialize/tests/builds/55395 and https://buildkite.com/materialize/nightlies/builds/2348
One way of checking if we are unexpectedly not using the new grpc system would be to make it panic in testing when it fails. That would tell us quickly if sqllogictest/testdrive/etc are correctly wired up.

def- commented:
Apparently all the cloudtests have problems with startup? https://buildkite.com/materialize/tests/builds/55395#01880f69-4c09-4a18-857f-5b2c26de5bf5

def- reviewed (changes requested):
Apparently all the cloudtests have problems with startup? https://buildkite.com/materialize/tests/builds/55395#01880f69-4c09-4a18-857f-5b2c26de5bf5
Seems related to this change

def- commented on misc/python/materialize/cloudtest/k8s/environmentd.py:
```diff
@@ -148,6 +148,8 @@ def generate_stateful_set(self) -> V1StatefulSet:
             "--storage-stash-url=postgres://root@cockroach.default:26257?options=--search_path=storage",
             "--internal-sql-listen-addr=0.0.0.0:6877",
+            "--internal-persist-pubsub-listen-addr=0.0.0.0:6879",
+            "--persist-pubsub-service-url=http://persist-pubsub",
```
Seems to come from here.
```suggestion
            "--persist-pubsub-url=http://persist-pubsub",
```

pH14 replied:
whoops! missed that when I did the last renaming, thanks for the pointer

def- commented on misc/python/materialize/cloudtest/k8s/environmentd.py:
```diff
@@ -148,6 +148,8 @@ def generate_stateful_set(self) -> V1StatefulSet:
             "--adapter-stash-url=postgres://root@cockroach.default:26257?options=--search_path=adapter",
             "--storage-stash-url=postgres://root@cockroach.default:26257?options=--search_path=storage",
             "--internal-sql-listen-addr=0.0.0.0:6877",
+            "--internal-persist-pubsub-listen-addr=0.0.0.0:6879",
```
```
[pod/environmentd-0/environmentd] error: Found argument '--internal-persist-pubsub-listen-addr' which wasn't expected, or isn't valid in this context
[pod/environmentd-0/environmentd] 
[pod/environmentd-0/environmentd] 	Did you mean '--internal-sql-listen-addr'?
```

pH14 replied:
was this locally or in CI?

def- replied:
In CI: https://buildkite.com/materialize/nightlies/builds/2349#01880fb6-4bf8-49f9-a751-ee68f9bb328a
See [kubectl-get-logs.log](https://buildkite.com/organizations/materialize/pipelines/nightlies/builds/2349/jobs/01880fb6-4bf8-49f9-a751-ee68f9bb328a/artifacts/01880fc0-618e-41a1-a980-bc016aa40852)

pH14 replied:
I am so puzzled by this one, it works fine locally for me and the error output there doesn't even list the other `--persist-pubsub-url` param that's added in

```
environmentd --orchestrator <ORCHESTRATOR> --persist-blob-url <PERSIST_BLOB_URL> --persist-consensus-url <PERSIST_CONSENSUS_URL> --storage-stash-url <POSTGRES_URL> --adapter-stash-url <POSTGRES_URL> --environment-id <<CLOUD>-<REGION>-<ORG-ID>-<ORDINAL>> --sql-listen-addr <HOST:PORT> --http-listen-addr <HOST:PORT> --availability-zone <AVAILABILITY_ZONE> --aws-account-id <AWS_ACCOUNT_ID> --aws-external-id-prefix <ID> --announce-egress-ip <ANNOUNCE_EGRESS_IP>... --orchestrator-kubernetes-image-pull-policy <ORCHESTRATOR_KUBERNETES_IMAGE_PULL_POLICY> --internal-sql-listen-addr <HOST:PORT>
```

pH14 replied:
that one is part of the `main` entrypoint to environmentd https://github.com/MaterializeInc/materialize/pull/18981/files#diff-18db540d7592188856fb9af92ebb418ad154a7b8c04e88c798f33a99b530c0a6R221-R232

def- replied:
internal-sql-listen-addr comes from `src/environmentd/src/bin/environmentd/main.rs`: 
```
    #[clap(
        long,
        value_name = "HOST:PORT",
        env = "INTERNAL_SQL_LISTEN_ADDR",
        default_value = "127.0.0.1:6877"
    )]
    internal_sql_listen_addr: SocketAddr,
```

pH14 replied:
Cloudtest is passing for the current build in CI: https://buildkite.com/materialize/tests/builds/55421#01880fc9-cedf-4efc-a6c0-1d81fc2d246a

pH14 replied:
Pretty puzzled by that error, I didn't see anything odd in your commits but it looks like that envd was running without both new params

pH14 replied:
Same nightly "Platform upgrade checks for Cloudtest/k8s" is failing on this build, but if you check the `kubectl describe all` logs the `environmentd` image running on the pods isn't the same as the one from the build... I think we might have a bug in the test harness

pH14 replied:
OH it's because the test is trying to run an upgrade compatibility check with the last release, but is using the new version's args. @def- let me know if you have suggestions here!

danhhz replied:
👍 trying that now in nightlies 🤞

danhhz replied:
that fixed it! @def-, paul had to take off for the day so i'm taking this over, but still ramping up on the context. what else would we like to do before getting this merged?

def- replied:
Great, my final step will be the same test with the flag on: https://buildkite.com/materialize/nightlies/builds/2357
If this is green I'm out of ideas for now, so good from my side.

def- replied:
The upgrade test is green now.

danhhz replied:
the other test you kicked off in that run failed, but looks like it's the same errors we're seeing in main

danhhz commented:
\o/ thanks for the help!

pH14 commented:
Thank you @def- for the QA assist + @danhhz for the final push here! 🙇
